### ISSUE TITLE
Forward hook called twice in detection model

### ISSUE THREAD

**Issue by wuhanshuo** [ultralytics/ultralytics#15757]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am trying to figure out how the yolov8 detection model is forwarded and get outputs from some layers.

I have this code block. 

```python
from ultralytics import YOLO

model = YOLO("yolov8n.pt").to(device)
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model.to(device)

img = torch.ones(1, 3, 640, 640).to(device)

def hook_fn(module, input, output):
    print("Hook is called!!")
    print(f"Input shape: {input[0].shape}")
    print(input)
    print(f"Output shape: {output.shape}")
    print(output)
    
layer = dict(model.named_modules())['model.model.0' ]
layer.register_forward_hook(hook_fn)

result = model.predict(img)
```

However I found the hook function is called twice, and each of them has different inputs. I believe the second one should be the correct input and output. why is this happening?

```

Hook is called!!!
Input shape: torch.Size([1, 16, 320, 320])
Input is: tensor([[23.3094,  1.6369,  1.7137,  ...,  2.3833,  0.6308,  2.6923],
        [26.9352,  0.9784,  0.9928,  ...,  1.2832,  0.8253,  1.8542],
        [26.7680,  1.0293,  0.1902,  ...,  1.0420,  1.3235,  1.1998],
        ...,
        [41.2127,  0.9562,  0.9562,  ...,  0.9562,  0.9562,  0.9562],
        [41.2127,  0.9562,  0.9562,  ...,  0.9562,  0.9562,  0.9562],
        [41.2127,  0.9562,  0.9562,  ...,  0.9562,  0.9562,  0.9562]], device='cuda:0')
Output shape: torch.Size([1, 32, 160, 160])
Output is: tensor([ 5.0925e+00, -8.0617e-05, -3.9494e-05, -7.4317e-05, -1.7077e-05, -9.5222e-05, -1.6611e-02, -3.2747e-04, -1.6107e-04, -4.8636e-05, -7.2875e-05, -1.2313e-04, -4.4289e-05, -2.8737e-05, -2.4214e-04, -6.3587e-04, -1.8644e-04, -5.5618e-04, -4.4560e-05, -2.2920e-04, -2.7716e-04, -8.3501e-05, -3.3116e-04, -3.2621e-04,
        -1.0800e-04, -7.5282e-04, -6.7144e-04, -1.7065e-04, -6.1899e-05, -1.2346e-05, -6.9699e-04, -7.1167e-03, -2.4942e-05, -1.4241e-05, -1.6079e-04, -6.4495e-04, -1.1928e-04, -8.2094e-05, -1.3497e-04, -5.9445e-05, -1.2968e-02, -2.4547e-04, -4.1662e-05, -1.3548e-04, -2.3282e-04, -2.7006e-05, -3.0753e-05, -1.2183e-04,
        -1.6160e-04, -6.3365e-05, -1.5308e-04, -8.0815e-05, -1.4050e-03, -1.4175e-03, -3.0110e-05, -2.4988e-04, -2.6437e-04, -1.8318e-04, -3.6555e-05, -1.6371e-04, -1.0320e-04, -2.4558e-04, -2.0110e-04, -6.0459e-04, -2.1267e-03, -4.3231e-04, -4.3378e-04, -8.0681e-04, -9.6905e-05, -1.8447e-04, -1.2139e-05, -2.1884e-04,
        -3.2534e-04, -6.3400e-05, -1.0835e-04, -2.2676e-04, -7.8086e-04, -1.8096e-05, -1.8083e-05, -2.1767e-04, -1.1412e-03, -9.8612e-06, -4.4332e-05, -5.1531e-05, -1.2874e-04, -3.6022e-04, -1.1172e-04, -4.9990e-05, -1.4992e-04, -2.0349e-04, -5.7087e-05, -2.0960e-04, -2.3030e-04, -6.9860e-04, -9.9053e-04, -2.8674e-04,
        -1.0118e-04, -3.5657e-03, -3.4973e-05, -1.6262e-04, -2.8795e-04, -1.2536e-04, -4.7791e-04, -1.3631e-04, -4.3932e-04, -1.2483e-05, -2.3044e-04, -7.9257e-04, -6.8666e-06, -6.6232e-05, -9.7919e-05, -1.0638e-04, -2.4183e-04, -6.8068e-05, -1.3805e-04, -1.2547e-05, -2.2364e-05, -1.4383e-04, -5.4983e-04, -4.6671e-04,
        -1.0860e-04, -4.2900e-04, -8.1124e-05, -9.1665e-06, -5.1373e-04, -2.6099e-04, -2.4096e-04, -6.9116e-06, -8.2470e-05, -2.5620e-04, -3.5840e-03, -7.8769e-05, -1.2121e-04, -1.0198e-04, -6.6673e-05, -3.1530e-04, -4.2592e-04, -4.2448e-05, -1.4407e-05, -3.7122e-05, -9.1933e-05, -2.3950e-05, -2.6347e-04, -1.4667e-04,
        -2.4077e-04, -4.7624e-04, -8.0247e-05, -1.7163e-04, -7.1525e-05, -7.2992e-05, -3.7965e-04, -7.9402e-04, -7.2533e-05, -7.8063e-05, -1.8375e-04, -2.2385e-03, -3.3140e-04, -5.4654e-05, -1.4494e-03, -1.5985e-02], device='cuda:0')
Hook is called!!!
Input shape: torch.Size([1, 16, 320, 320])
Input is: tensor([[36.4291,  1.6872,  1.6872,  ...,  1.6872,  1.6872,  1.6872],
        [41.2127,  0.9562,  0.9562,  ...,  0.9562,  0.9562,  0.9562],
        [41.2127,  0.9562,  0.9562,  ...,  0.9562,  0.9562,  0.9562],
        ...,
        [41.2127,  0.9562,  0.9562,  ...,  0.9562,  0.9562,  0.9562],
        [41.2127,  0.9562,  0.9562,  ...,  0.9562,  0.9562,  0.9562],
        [41.2127,  0.9562,  0.9562,  ...,  0.9562,  0.9562,  0.9562]], device='cuda:0')
Output shape: torch.Size([1, 32, 160, 160])
Output is: tensor([ 3.0506e+00, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06,
        -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06,
        -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06,
        -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06,
        -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06,
        -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06,
        -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06, -3.6873e-06], device='cuda:0')
0: 640x640 (no detections), 20.5ms
Speed: 0.0ms preprocess, 20.5ms inference, 2.0ms postprocess per image at shape (1, 3, 640, 640)
```

**Comment by Y-T-G** [comment]
It only happens in the first predict call because it sets up the predictor and does warmup:

https://github.com/ultralytics/ultralytics/blob/0c6e1e16bd0778ac7a6f86eb794e3c33257e2028/ultralytics/engine/predictor.py#L234

**Comment by wuhanshuo** [comment]
> It only happens in the first predict call because it sets up the predictor and does warmup:
> 
> https://github.com/ultralytics/ultralytics/blob/0c6e1e16bd0778ac7a6f86eb794e3c33257e2028/ultralytics/engine/predictor.py#L234

Thanks! I see it now

**Comment by glenn-jocher** [comment]
@wuhanshuo you're welcome! If you have any more questions or need further assistance, feel free to ask.
